src/gitlab.js

```diff
@@ -172,6 +218,30 @@ export default class VerdaccioGitLab implements IPluginAuth {
   allow_publish(user: RemoteUser, _package: VerdaccioGitlabPackageAccess, cb: Callback) {
     if (!_package.gitlab) return cb(null, false);
 
+    if (_package.gitlabUserGroupWhitelist) {
+      const cachedUserGroups = this._getCachedUserGroups(user.name);
+      if (!cachedUserGroups) {
+        this.logger.debug(`[gitlab] no user found in cache for this username: ${user.name}`);
+        return cb(httperror[401]('publish denied, user found'));
+      }
+
+      let userGroups = cachedUserGroups.publish;
+
+      this.logger.trace(`[gitlab] publish: checking group whitelist for user: ${user.name || ''} and package: ${_package.name}`);
+
+      const gitlabUserGroupWhitelist = normalizeYamlSetting(_package.gitlabUserGroupWhitelist);
+
+      for (let whitelistedGroup of gitlabUserGroupWhitelist) {
+        if (userGroups.indexOf(whitelistedGroup) !== -1) {
+          this.logger.debug(`[gitlab] user: ${user.name || ''} allowed to publish package: ${_package.name} based on group whitelist ${whitelistedGroup}`);
+          return cb(null, true);
+        }
+      }
+
+      this.logger.debug(`[gitlab] deny publish to package: ${_package.name}`);
+      return cb(httperror[401]('publish denied, user is not member of any of the whitelisted groups'));
+    }
+    
```

Comment:
Same check as for the `allow_access` above : Check if the user belongs to one of the whitelisted groups.